In [1]:
import numpy as np # Import package numpy untuk perhitungan matematis
from tabulate import tabulate # Import package tabulate untuk membuat tabel
from sympy import * # Import package sympy untuk variabel pada formula

Program ini adalah program untuk meng-*import* suatu module ke dalam program.

Idenya adalah meng-*import* module dan/atau fungsi dari module yang akan digunakan pada program.

Algoritmanya:
1. Meng-*import* module `numpy` dan panggil sebagai `np`.
1. Meng-*import* `tabulate` dari module `tabulate`
1. Meng-*import* seluruh fungsi-fungsi dari module `sympy`

## **Turunan Numerik**
Di bawah ini, $h \neq 0$
###  a) Three point formula
Endpoint: $f'(x_0) \approx \frac{1}{2h}[-3f(x_0)+4f(x_0+h)-f(x_0+2h)] $

Midpoint: $f'(x_0) \approx \frac{1}{2h}[f(x_0+h)-f(x_0-h)]$

### b) Five point formula
Endpoint: $f'(x_0) ≈ \frac{1}{12h}[-25f(x_0)+48f(x_0+h)-36f(x_0+2h)+16f(x_0+3h)-3f(x_0+4h)]$

Midpoint: $f'(x_0) \approx \frac{1}{12h}[f(x_0-2h)-8f(x_0-h)+8f(x_0+h)-f(x_0+2h)]$

In [2]:
def TPEP(x,h): # Three Point End Point
    return (1/(2*h))*(-3*f(x)+4*f(x+h)-f(x+2*h))

def TPMP(x,h): # Three Point Mid Point
    return (1/(2*h))*(f(x+h)-f(x-h))

def FPEP(x,h): # Five Point End Point
    return (1/(12*h))*(-25*f(x)+48*f(x+h)-36*f(x+2*h)+16*f(x+3*h)-3*f(x+4*h))

def FPMP(x,h): # Five Point Mid Point
    return (1/(12*h))*(f(x-2*h)-8*f(x-h)+8*f(x+h)-f(x+2*h))

## **Integral Numerik**

Pada dasarnya, ada dua metode numerik untuk integral, yaitu:

###  a) Trapezoid
$$ \int_{a}^{b} f(x) dx \approx \frac{h}{2}[f(x_0)+f(x_1)]$$

dengan $x_0 = a, x_1 = b, h = b-a$

### b) Simpson
$$ \int_{a}^{b} f(x) dx \approx \frac{h}{3}[f(x_0)+4f(x_1)+f(x_2)]$$

dengan $x_0 = a, x_1 = a + h, x_2 = b, h = \frac{(b-a)}{2}$

In [3]:
def Trapezoid(a,b): # Metode Trapesium
    h = b-a # hitung h dari selisih b dan a
    return (h/2)*(f(a)+f(b))

def Simpson(a,b): # Metode Simpson
    h = (b-a)/2 # hitung h dari selisih b dan a yang dibagi 2
    return (h/3)*(f(a)+4*f(a+h)+f(b))

## **Integral Numerik Komposit**

Integrasi numerik seringkali menghasilkan hasil yang tidak akurat untuk interval yang cukup panjang. Untuk mengatasi masalah tersebut, interval tersebut dibagi menjadi bagian sama besar, kemudian lakukan integrasi numerik (Trapezoidal/Simpson) pada setiap bagian. Jika menggunakan trapezoidal, maka metode ini disebut *composite trapezoidal*/trapezoid komposit dan jika menggunakan Simpson, maka metode ini disebut *composite Simpson*/Simpson komposit.

###  a) Trapezoid Komposit / *Composite Trapezoidal*
$$ \int_{a}^{b} f(x) dx \approx \frac{h}{2}[f(a)+ 2\sum_{j=1}^{n-1}f(x_j) +f(b)]$$

dengan $n \in {Z}^+, h=\frac{(b-a)}{n}, x_j = a + jh, j = 0,1, ... , n$


### b) Simpson Komposit / *Composite Simpson*
$$ \int_{a}^{b} f(x) dx \approx \frac{h}{3}[f(a)+2\sum_{j=1}^{(n/2)-1}f(x_{2j})+4\sum_{j=1}^{n/2}f(x_{2j}-1)+f(b)]$$

dengan $n = 2m,m\in Z^{+}, h=\frac{(b-a)}{n}, x_j=a+jh, j = 0,1, ... ,n$

In [4]:
def CompTrapezoid(a,b,n): # Trapezoid Komposit
    h = (b-a)/n # mencari nilai h dari selisih a dan b yang dibagi partisinya (n)
    sumtrap = 0 # Inisiasi nilai sumasi trapezoid = 0
    for j in range (1,n): # proses pengisian sumasi
        sumtrap += f(a+j*h)
    return (h/2)*(f(a)+2*sumtrap+f(b))

def CompSimpson(a,b,n): # Simpson Komposit
    if n % 2 == 1: # Jika partisi ganjil
        return "Jumlah partisi harus genap"
    elif n % 2 == 0: # Jika partisi genap
        h = (b-a)/n # mencari nilai h dari selisih a dan b yang dibagi partisinya (n)
        sum1, sum2 = 0, 0 # Inisiasi nilai sumasi 1 dan 2 sama dengan 0
        for j in range (1, int(n/2)): # proses pengisian sumasi
            sum1 += f(a+2*j*h)
            sum2 += f(a+(2*j-1)*h)
        sum2 += f(b-h) # selesai pengisian, sumasi ke-2 ditambah lagi dengan f(b-h)
        return (h/3)*(f(a)+2*sum1 + 4*sum2 + f(b))

## **Code**

In [5]:
x = symbols('x') # Simpan variabel x dalam bentuk symbol

while True: # loop diulang terus
    bonus = str(input("Ini buat soal bonus: (Y/N) ")).lower()
    if bonus not in ["y","n"]: # Jika input bukan y/n
        print("Anda salah input, silakan ulang.") # User diminta mengulang
    else: # Jika input y/n
        break # Program keluar dari loop
    
if bonus == "y": # Jika user memilih bonus
    npm = input("Masukan NPM: ") # Simpan input NPM user dalam variabel 'npm'

    npm_a = npm[-3] # Simpan digit ke-3 terakhir
    npm_b = npm[-2] # Simpan digit ke-2 terakhir
    npm_c = npm[-1] # Simpan digit terakhir

    print("Nilai variabel a:",npm_a) # Cetak digit ke-3 terakhir sebagai nilai variabel a
    print("Nilai variabel b:",npm_b) # Cetak digit ke-2 terakhir sebagai nilai variabel b
    print("Nilai variabel c:",npm_c) # Cetak digit terakhir sebagai nilai variabel c

    a, b, c = symbols('a b c') # Simpan variabel a, b, dan c dalam bentuk symbol

    fungsi = eval(input("Masukan formula fungsi yang akan dicari nilai turunan/integralnya: ")) # Menyimpan hasil evaluasi input formula fungsi dari user
    formula = str(fungsi.subs([(a, npm_a), (b, npm_b), (c, npm_c)])) # Substitusi nilai a, b, dan c pada fungsi input user dan paksa menjadi bentuk string

    print("Formula yang digunakan:",formula) # Cetak fungsi yang telah disubstitusi
    
elif bonus == "n": # Jika user memilih tidak bonus
    formula = input("Masukan formula fungsi yang akan dicari nilai turunan/integralnya: ") # Simpan input formula fungsi dari user

def f(x): # Fungsi untuk mengevaluasi formula yang sudah diinput user 
    return eval(formula) # Evaluasi formula hasil input user

program = [[1,"Turunan Numerik", 'Mengaproksimasi nilai turunan fungsi pada suatu titik'],\
           [2,"Integral Numerik",'Mengaproksimasi nilai integral pada interval [a,b]'],\
           [3,"Integral Numerik Komposit",'Integral numerik dengan partisi']] # Buat list yang akan dicetak dalam bentuk tabel

print(tabulate(program, headers = ['Kode', 'Program', 'Definisi'], tablefmt='pretty')) # Cetak tabel dengan list yang telah dibuat

while True: # Untuk mengulang program secara terus-menerus
    print("\nSelamat datang di program Turunan, Integral, dan Integral Komposit") # Cetak informasi selamat datang
    jenis = int(input("Pilih jenis program ('1','2','3'): ")) # Untuk menyimpan jenis program yang dipilih oleh user, simpan dalam bentuk integer
    
    # Metode Turunan Numerik
    if jenis == 1: # Jika user memilih program Turunan Numerik
        print(program[0][1]) # Cetak informasi jenis program yang dilakukan
        
        x0 = eval(input("Masukan titik di mana nilai turunannya akan diaproksimasi: ")) # Simpan hasil evaluasi input titik nilai aproksimasi turunan
        h = eval(input("Masukan besar stepsize: ")) # Simpan hasil evaluasi stepsize
        
        while True: # Proses pemilihan dilakukan secara looping terus untuk menghindari error pada program
            output = str(input("Output berupa tabel? (Y/N): ")).lower() # Simpan jenis output yang diinginkan user dalam bentuk string dan paksa menjadi lowercase
            if output not in ["y","n"]: # jika input bukan y/n
                print("Anda salah input, silakan ulang.") # user diminta mengulang
            else: # jika input y/n
                break # program keluar dari loop
    
        print("\nTurunan fungsi f di xo = {0} adalah:".format(x0)) # Cetak hasil turunan fungsi dengan teks
        
        if output == "y": # Jika user memilih output tabel
            turunan_num = [["TPEP",TPEP(x0,h)],\
                           ["TPMP",TPMP(x0,h)],\
                           ["FPEP",FPEP(x0,h)],\
                           ["FPMP",FPMP(x0,h)]] # Buat list untuk tabel yang akan dicetak dengan tiap metode serta hasilnya
        
            print(tabulate(turunan_num, headers = ['Metode Turunan Numerik','Aproksimasi'], tablefmt='pretty')) # Cetak tabel hasil turunan dan perhitungan turunannya
            
        elif output == "n": # Jika user tidak memilih output tabel
            print("f'({0}) = {1} (Three point endpoints)".format(x0,TPEP(x0,h))) # Cetak TPEP
            print("f'({0}) = {1} (Three point midpoints)".format(x0,TPMP(x0,h))) # Cetak TPMP
            print("f'({0}) = {1} (Five point endpoints)".format(x0,FPEP(x0,h))) # Cetak FPEP
            print("f'({0}) = {1} (Five point midpoints)".format(x0,FPMP(x0,h))) # Cetak FPMP
    
    # Metode Integral Numerik
    elif jenis == 2: # jIka user memilih program Integral Numerik
        print(program[1][1]) # Cetak informasi jenis program yang dilakukan
        
        low_bound = eval(input('Masukan batas bawah integral: ')) # Untuk menyimpan batas bawah yang akan dihitung
        up_bound = eval(input('Masukan batas atas integral: ')) # Untuk menyimpan batas atas yang akan dihitung
        
        while True: # Proses pemilihan dilakukan secara looping terus untuk menghindari error pada program
            output = str(input("Output berupa tabel? (Y/N): ")).lower() # Simpan jenis output yang diinginkan user dalam bentuk string dan paksa menjadi lowercase
            if output not in ["y","n"]: # Jika input bukan y/n
                print("Anda salah input, silakan ulang.") # User diminta mengulang
            else: # Jika input y/n
                break # Program keluar dari loop
    
        print("\nHasil integral dari fungsi f(x) = {0} pada interval [{1},{2}] adalah: ".format(formula, low_bound, up_bound))
        
        if output == "y": # Jika user memilih y
            int_num = [["Trapezoid",Trapezoid(low_bound, up_bound)],\
                       ["Simpson",Simpson(low_bound, up_bound)]] # Buat list untuk tabel yang akan dicetak dengan tiap metode serta hasilnya
        
            print(tabulate(int_num, headers = ['Metode Integral Numerik','Aproksimasi'], tablefmt='pretty')) # Cetak tabel hasil integrasi dan perhitungan integrasinya
            
        elif output == "n": # Jika user memilih n
            print("Trapezoid = {0}".format(Trapezoid(low_bound, up_bound))) # Cetak hasil metode Trapezoid
            print("Simpson = {0}".format(Simpson(low_bound, up_bound))) # Cetak hasil metode Simpson

    # Metode Integral Numerik Komposit
    elif jenis == 3: # jIka user memilih program Integral Numerik Komposit
        print(program[2][1]) # Cetak informasi jenis program yang dilakukan
        
        low_bound = eval(input("Masukan batas bawah: ")) # Untuk menyimpan batas bawah yang akan dihitung
        up_bound = eval(input("Masukan batas atas: ")) # Untuk menyimpan batas atas yang akan dihitung
        partisi = eval(input("Masukan jumlah partisi (n): ")) # Untuk menyimpan banyaknya partisi yang akan dilakukan
        
        while True: # Proses pemilihan dilakukan secara looping terus untuk menghindari error pada program
            output = str(input("Output berupa tabel? (Y/N): ")).lower() # Simpan jenis output yang diinginkan user dalam bentuk string dan paksa menjadi lowercase
            if output not in ["y","n"]: # Jika input bukan y/n
                print("Anda salah input, silakan ulang.") # User diminta mengulang
            else: # Jika input y/n
                break # Program keluar dari loop
    
        print("\nHasil integral numerik komposit dari fungsi f(x) = {0} pada interval [{1},{2}] dengan {3} partisi adalah:".format(formula,low_bound,up_bound,partisi)) # Cetak informasi untuk perhitungan yang telah diinput user
        
        if output == "y": # Jika user memilih output tabel
            intcomp_num = [["Composite Trapezoid",CompTrapezoid(low_bound, up_bound, partisi)],\
                           ["Composite Simpson",CompSimpson(low_bound, up_bound, partisi)]] # Buat list untuk tabel yang akan dicetak dengan tiap metode serta hasilnya
        
            print(tabulate(intcomp_num, headers = ['Metode Integral Numerik','Aproksimasi'], tablefmt='pretty')) # Cetak tabel hasil integrasi dan perhitungan integrasinya
            
        elif output == "n": # Jika user tidak memilih output tabel
            print("Composite trapezoid dengan {0} partisi = {1:.5f}".format(partisi,CompTrapezoid(low_bound, up_bound, partisi))) # Cetak hasil metode Trapezoid Komposit dengan informasi jumlah partisinya
            print("Simpson trapezoid dengan {0} partisi = {1:.5f}".format(partisi,CompSimpson(low_bound, up_bound, partisi))) # Cetak hasil metode Simpson Komposit dengan informasi jumlah partisinya

    while True: # Proses pemilihan dilakukan secara looping terus untuk menghindari error pada program
        ulang = str(input("Mau menghitung ulang turunan/integral? (Y/N): ")).lower() # Simpan input yang diinginkan user dalam bentuk string dan paksa menjadi lowercase
        if ulang not in ["y","n"]: # Jika input bukan y/n
            print("Anda salah input, silakan ulang.") # User diminta mengulang
        else: # Jika input y/n
            break # Program keluar dari loop
    
    print("\n============================================================") # Cetak pemisah antar perulangan program
    
    if ulang == "n": # Jika user memilih n
        break # Program keluar dan berhenti

print("\n~Terima Kasih~") # Cetak pernyataan "Terima Kasih"

Ini buat soal bonus: (Y/N) Y
Masukan NPM: 2106725034
Nilai variabel a: 0
Nilai variabel b: 3
Nilai variabel c: 4
Masukan formula fungsi yang akan dicari nilai turunan/integralnya: x**5 + 9*x**4 - a*x**3 + b*x**2 - c*x + 5
Formula yang digunakan: x**5 + 9*x**4 + 3*x**2 - 4*x + 5
+------+---------------------------+-------------------------------------------------------+
| Kode |          Program          |                       Definisi                        |
+------+---------------------------+-------------------------------------------------------+
|  1   |      Turunan Numerik      | Mengaproksimasi nilai turunan fungsi pada suatu titik |
|  2   |     Integral Numerik      |  Mengaproksimasi nilai integral pada interval [a,b]   |
|  3   | Integral Numerik Komposit |            Integral numerik dengan partisi            |
+------+---------------------------+-------------------------------------------------------+

Selamat datang di program Turunan, Integral, dan Integral Komposit
Pi

Sebuah fungsi $f(x)$ didefinisikan sebagai $x^5+9x^4-ax^3+bx^2-cx+5=0$ dimana $a$, $b$, $c$ adalah tiga digit terakhir berturut-turut dari NPM. Maka, untuk NPM saya, $2106725034$, fungsi yang digunakan adalah $x^5+9x^4+3x^2-4x+5=0$

Program ini adalah program untuk menghitung turunan dan/atau integral dengan input fungsi dari *user* dan mencetak hasil perhitungannya sesuai dengan keinginan *user*.

Idenya adalah membuat program berulang yang meminta input formula atau fungsi dari *user*. Kemudian, meminta *user* untuk memilih jenis program serta jenis output. Lalu, dari jenis output tersebut, akan dicetak output yang sesuai dengan pilihan *user*.

* Jika user memilih diturunkan, tampilkan
    * TPMP
    * TPEP
    * FPMP
    * FPEP
* Jika user memilih diintegralkan, tampilkan
    * Metode Trapezoid
    * Metode Simpson
* Jika user memilih diintegralkan komposit, tampilkan
    * Metode Trapezoid Komposit
    * Metode Simpson Komposit

Algoritmanya:
1. Membuat variabel `x` dan mengisinya dalam bentuk `symbols` dengan bantuan *module* `sympy`.
1. Membuat *loop* untuk input permintaan soal 'bonus' atau 'tidak'. Jika inputnya tidak sesuai, *user* akan diminta input ulang hingga inputnya sesuai.
    1. Jika *user* memilih Y (bonus)
        1. *User* diminta untuk mengisi NPM.
        1. Menyimpan tiga digit terakhir dari NPM di tiga variabel yang berbeda, `npm_a`, `npm_b`, dan `npm_c`.
        1. Cetak informasi masing-masing variabel `a`, `b`, dan `c`  dari `npm_a`, `npm_b`, dan `npm_c`.
        1. Buat variabel `a`, `b`, dan `c` dan mengisinya dalam bentuk `symbols` dengan bantuan *module* `sympy`.
        1. Meminta *user* memasukkan formula fungsi.
        1. Substitusi nilai `a`, `b`, dan `c` dari tiga digit terakhir NPM ke dalam fungsi yang telah di-*input* *user*.
        1. Cetak hasil substitusi pada formula tersebut ke pada *user*.
    1. Jika *user* memilih N (tidak bonus), *user* diminta untuk langsung memasukkan formula fungsi.
1. Buat fungsi `f(x)` yang me-*return* nilai evaluasi `formula`.
1. Buat variabel `program` dengan isi *list* informasi program.
1. Cetak tabel dari list `program` dan buat `headers` juga dengan bantuan `tabulate`.
1. Buat *loop* operasi program Turunan, Integral, dan Integral Komposit.
    1. Cetak informasi selamat datang pada *user*.
    1. Meminta *user* untuk meng-*input* jenis program yang akan dipilih dan simpan ke dalam variabel `jenis`
        1. Jika *user* memilih jenis program Turunan Numerik `1`
            1. Cetak informasi jenis program yang akan dilakukan
            1. Meminta *user* untuk mengisi titik aproksimasi dan simpan dalam variabel `x0`
            1. Meminta *user* untuk mengisi besar *stepsize* dan simpan dalam variabel `h`
            1. Membuat *loop* untuk input permintaan output 'tabel' atau 'tidak'. Jika inputnya tidak sesuai, *user* akan diminta input ulang hingga inputnya sesuai. Jika sesuai, maka program akan keluar dari *loop*
            1. Cetak informasi turunan f di xo dari variabel `x0` dengan fungsi `.format()`
            1. Jika *user* memiliih output tabel `Y`
                1. Buat *list* pada variabel `turunan_num` yang berisi informasi program (TPMP, TPEP, FPMP, FPEP) dan pemanggilan program yang telah didefinisi dengan variabel yang sudah diisi *user* juga (`x0`, `h`).
                1. Cetak tabel dari list `turunan_num` dan buat `headers` juga dengan bantuan `tabulate`.
            1. Jika *user* memilih output tidak tabel `N`
                1. Cetak masing-masing informasi program (TPMP, TPEP, FPMP, FPEP) dan hasil perhitungannya dengan bantuan `.format()`
        1. Jika *user* memilih jenis program Integral Numerik `2`
            1. Cetak informasi jenis program yang akan dilakukan
            1. Meminta *user* untuk mengisi batas bawah dan simpan dalam variabel `low_bound`
            1. Meminta *user* untuk mengisi batas atas dan simpan dalam variabel `up_bound`
            1. Membuat *loop* untuk input permintaan output 'tabel' atau 'tidak'. Jika inputnya tidak sesuai, *user* akan diminta input ulang hingga inputnya sesuai. Jika sesuai, maka program akan keluar dari *loop*
            1. Cetak informasi formula (`formula`), batas bawah (`low_bound`), dan batas atas (`up_bound`) dengan fungsi `.format()`
            1. Jika *user* memiliih output tabel `Y`
                1. Buat *list* pada variabel `int_num` yang berisi informasi program (Trapezoid, Simpson) dan pemanggilan program yang telah didefinisi dengan variabel yang sudah diisi *user* juga (`low_bound`, `up_bound`).
                1. Cetak tabel dari list `int_num` dan buat `headers` juga dengan bantuan `tabulate`.
            1. Jika *user* memilih output tidak tabel `N`
                1. Cetak masing-masing informasi program dan hasil perhitungannya dengan bantuan `.format()`
        1. Jika *user* memilih jenis program Integral Numerik Komposit `3`
            1. Cetak informasi jenis program yang akan dilakukan
            1. Meminta *user* untuk mengisi batas bawah dan simpan dalam variabel `low_bound`
            1. Meminta *user* untuk mengisi batas atas dan simpan dalam variabel `up_bound`
            1. Meminta *user* untuk mengisi berapa partisi yang akan dilakukan dan simpan dalam variabel `partisi`
            1. Membuat *loop* untuk input permintaan output 'tabel' atau 'tidak'. Jika inputnya tidak sesuai, *user* akan diminta input ulang hingga inputnya sesuai. Jika sesuai, maka program akan keluar dari *loop*
            1. Cetak informasi formula (`formula`), batas bawah (`low_bound`), batas atas (`up_bound`), dan banyaknya partisi (`partisi`) dengan fungsi `.format()`
            1. Jika *user* memiliih output tabel `Y`
                1. Buat *list* pada variabel `intcomp_num` yang berisi informasi program (Trapezoid Komposit, Simpson Komposit) dan pemanggilan program yang telah didefinisi dengan variabel yang sudah diisi *user* juga (`low_bound`, `up_bound`, `partisi`).
                1. Cetak tabel dari list `intcomp_num` dan buat `headers` juga dengan bantuan `tabulate`.
            1. Jika *user* memilih output tidak tabel `N`
                1. Cetak masing-masing informasi program dan hasil perhitungannya dengan bantuan `.format()`
    1. Membuat *loop* untuk *input* permintaan 'hitung ulang' atau 'tidak'. Jika inputnya tidak sesuai, *user* akan diminta input ulang hingga inputnya sesuai. Jika sesuai, maka program akan keluar dari *loop*
    1. Jika *input user* adalah tidak (`N`), program akan keluar dari `loop`.
1. Cetak pesan "terima kasih" pada *user*